src/ansi.rs

```diff
@@ -0,0 +1,1933 @@
+//! ANSI Terminal Stream Parsing.
+
+use alloc::borrow::ToOwned;
+use alloc::string::{String, ToString};
+use alloc::vec::Vec;
+
+use core::convert::TryFrom;
+use core::fmt::Write;
+use core::time::Duration;
+use core::{iter, str};
+
+use log::{debug, trace};
+
+#[cfg(feature = "ansi-serde")]
+use serde::{Deserialize, Serialize};
+
+use crate::{Params, ParamsIter};
+
+pub trait SyncHandler: Default {
+    /// Expiration time of the synchronized update.
+    fn update_timeout(&mut self, _: Option<Duration>);
+    fn pending_timeout(&self) -> bool;
+}
+
+#[cfg(not(feature = "no_std"))]
+pub type DefaultSyncHandler = StdSyncHandler;
+
+#[cfg(not(feature = "no_std"))]
+#[derive(Default)]
+pub struct StdSyncHandler {
+    timeout: Option<std::time::Instant>,
+}
+
+#[cfg(not(feature = "no_std"))]
+impl StdSyncHandler {
+    /// Synchronized update expiration time.
+    #[inline]
+    pub fn sync_timeout(&self) -> Option<std::time::Instant> {
+        self.timeout
+    }
+}
+
+#[cfg(not(feature = "no_std"))]
+impl SyncHandler for StdSyncHandler {
+    fn update_timeout(&mut self, duration: Option<Duration>) {
+        use std::time::Instant;
+        self.timeout = duration.map(|e| Instant::now() + e);
+    }
+
+    #[inline]
+    fn pending_timeout(&self) -> bool {
+        self.timeout.is_some()
+    }
+}
+
+#[cfg(feature = "no_std")]
+pub type DefaultSyncHandler = NullTimeProvider;
+
+/// Time provider which always returns 0.
+#[derive(Default)]
+pub struct NullTimeProvider;
+
+impl SyncHandler for NullTimeProvider {
+    fn update_timeout(&mut self, _: Option<Duration>) {}
+
+    fn pending_timeout(&self) -> bool {
+        false
+    }
+}
```

Comment:
We use master layout for definitions, in a way that definition comes after the use, with some priorities. `const` stuff should be above the rest, and the structure's will move to the bottom.

Reply:
Do the `impl`s be after the structure definition as-well? If I understood this correctly, the correct layout would be type aliases, structure definition and implementation pair and at the end will be the trait definition?

Reply:
```rust
impl Other {
    fn other_fn() -> Foo {
    }
}

struct Foo {
}
impl Foo {
}
impl Drop for Foo {
}

impl OtherTraits for Foo {
}
```

Reply:
fixed in https://github.com/alacritty/vte/pull/85/commits/994c055115c4f990f759737cf14bf8f37b213131